src/systemtest/resources/30daysFromUntilDate/expected/reposense_testrepo-Charlie_master/authorship.json

```diff
@@ -9926,7 +9926,7 @@
       {
         "lineNumber": 39,
         "author": {
-          "gitId": "charlesgoh"
+          "gitId": "-"
```

Comment:
this one seems to be wrong, please check that your solution is working; no discrepancy between results and fact before requesting for review.

Reply:
@eugenepeh 
I don't think this is wrong as this line was committed during `1st Nov 2017 10.35am GMT+8`.
<img width="755" alt="Screen Shot 2019-07-24 at 11 47 02 AM" src="https://user-images.githubusercontent.com/32864116/61763753-1c0c0800-ae09-11e9-926a-e6e1702c7029.png">

Whereas the test analysis period is from `1st Oct 2017` to `1st Nov 2017`. The current implementation of setting sinceDate and untilDate is getting the date at `12.00am`. Since `1st Nov 2017 10.35am` is after `1st Nov 2017 12.00 am`, the line should be set as an unknown author `-`.
<img width="864" alt="Screen Shot 2019-07-24 at 11 48 48 AM" src="https://user-images.githubusercontent.com/32864116/61763835-75743700-ae09-11e9-91ba-5c5879a0aae4.png">

Reply:
yes but the `"content": "image::Ui.png[width\u003d\"800\"]"` below is the one updated by charlesgoh, isn't it? So why is his gitId not being tagged to the content?

Reply:
Yes it is updated by charlesgoh. 

This PR basically sets any line's author to an unknown `-` when the commit date of that line is not within the since date and until date. 

The reason being is to prevent the situation as follow:
Since we are using the option `-w` in our git blame, users that moved a specific line without modifying the content will not be recognised as the author of that line, the previous author will be recognised as the author instead. But if the previous author's commit date for that particular line is before the since date, the code view will still show this line. As said by Prof Damith,

> The code should not be in the chart because Jun An didn't contribute the code during the analysis period. 

_Originally posted by @damithc in https://github.com/reposense/RepoSense/issues/785#issuecomment-507974771_

Hence setting these lines to an unknown `-` will prevent the line to be shown in the code view.

Reply:
Yes, but the content should match the date and the author as well. Because when you define an `untilDate`, RepoSense is supposed to check out to the latest commit before the date as well. So if charlesgoh is not suppose to be the author, the `"content": "image::Ui.png[width\u003d\"800\"]"` should not be the content as well.

Reply:
I'm still getting a grasp of this issue, but meanwhile in `GitRevList#getCommitHashUntilDate`:

![image](https://user-images.githubusercontent.com/35756209/61843920-293ffa00-aed0-11e9-9051-16061aad8ca2.png)

I want the resulting commit hash made **before 1 Nov 2017, 00:00hrs** but the resulting `revListCommand` to be used reflects a date-time to be **before 1 Nov 2017, 25:59:59**.

This results in the commit has `13762e5ed965557cd8bd90ad789ba280d095a91e` which falls on **1 Nov 2019, 21:59hrs SGT**.

In the UG:
![image](https://user-images.githubusercontent.com/35756209/61844054-b1260400-aed0-11e9-9b83-8dd85ec70e38.png)

It states that the analysis **excludes the end date** (i.e.: The until date should be **before** 1 Nov 2017, 00:00 hrs).


Now, tracing the source code: 
![image](https://user-images.githubusercontent.com/35756209/61844490-33fb8e80-aed2-11e9-812c-6c05ea9886ea.png)
![image](https://user-images.githubusercontent.com/35756209/61844591-889f0980-aed2-11e9-9b9f-a10ff75a7c8c.png)


The docs states `Checks out to the latest commit before {@code untilDate} in {@code branchName} branch if {@code untilDate} is not null.` But going into the method `getCommitHashUntilDate`, the docs states that `Returns the latest commit hash inclusive and until the end of the day of {@code date}.`.

Why is there a contradiction here?

Reply:
This was due to the result of several different opinions about the until date in the past. 
We ended with a conclusion that the until date should be included in the analysis, however I guess the docs were not updated to match that conclusion. Do open an issue to fix up on that. 

Anyway, regarding the result, it is still wrong in the sense where the content, author and date doesn't seems to fit together. Hence, @jamessspanggg you will still need to research on that.

Reply:
@eugenepeh for the example of charlesgoh that you gave above, the commit date was actually `11/7/2018 10.35a.m.` whereas our until date is `11/7/2018 12.00a.m.` If we set the until date to `11/7/2019 11.59p.m.`, the content will match the date and the author.

Reply:
Hmm ok. Will open up an issue on that.

But in any case, I think this until date issue also plays a part in whether the authorship should be credited to `charlesgoh` because a different hash was checked out as a result of the confusion as to whether untilDate is to be included or not.

As we thought that the until date should have been excluded in the analysis, let's say that I modify the `GIT_LOG_UNTIL_DATE_FORMAT` to take the date at 00:00hrs instead of 23:59hrs, and get the latest commit hash before 1 Nov 2017 00:00hrs. The commit hash returned should be `bde21cfaf0bfe306931b2670c56dadecf223512a`, which falls on 31/10/2017, 22:32hrs.

Here at [line 39 of their UG](https://github.com/reposense/testrepo-Charlie/blame/bde21cfaf0bfe306931b2670c56dadecf223512a/docs/UserGuide.adoc#L39), line 39 should be credited to `limmlingg` or `-` (depending on the author and date configs).


![image](https://user-images.githubusercontent.com/35756209/61884494-e82b0280-af2e-11e9-8cf9-24ada4f91a0d.png)
Here is the `30daysfromutildate` system test, where the left side shows the result when I were to check out to the hash that is before the until date **(exclusive)**, where the right side shows the current implementation result where RepoSense checks out to the hash that is the untilDate **inclusive**.

Reply:
@jamessspanggg 

Hmm, revlist and checkout should be using the same until date format, hence I still suspecting that somewhere is wrong.

Currently, going by what you mentioned:
Yes, the date matches the author, however the content still doesn't match. If it is supposed to be the content that is not authored by `charlesgoh` the width should be 790 and not 800.
Hence, there is still a mismatch between content and author even though, assuming that that's an issue with the dates.

Reply:
@eugenepeh revlist uses until date that is `11.59p.m.` even though the until date is `12.00a.m.`.
![Uploading Screen Shot 2019-07-25 at 11.07.04 PM.png…]()

The mismatch of content and author is due to the until date in the config not matching with the until used in realist.

Reply:
>The until date in the config was used to compare whether the commit date is within range of since date and until date.

I see, in that case, do put up a fix to make both synchronize to 23:59 of the until date.

Reply:
Ok. Will do so in a separate pr.

Reply:
@eugenepeh the current issue no longer exists and this PR is ready for review.